src/jrd/btr.cpp

```diff
@@ -371,7 +372,31 @@ USHORT BTR_all(thread_db* tdbb, jrd_rel* relation, IndexDescAlloc** csb_idx, Rel
 	for (USHORT i = 0; i < root->irt_count; i++)
 	{
 		if (BTR_description(tdbb, relation, root, &buffer[count], i))
-			count++;
+		{
+			// So, if someone started to delete index, but transaction is not committed
+			// DropIndexNode erased index record in RDB$INDICES, but btree
+			// holds the index in case of cancellation of transaction.
+			// To ensure that the index will not be selected on the next query,
+			// we check, that the index does not have a lock on the deletion. If index lock exists,
+			// it means that the index is deleted and no need to use it.
+			// If the transaction rolled back, index name is restored in table
+			// and the index is used in the usual way.
+			bool skip = false;
+			Lock* lock = FB_NEW_RPT(*relation->rel_pool, 0) Lock(tdbb, sizeof(SLONG), LCK_idx_deletion);
+			lock->setKey((relation->rel_id << 16) | buffer[count].idx_id);
+			const SLONG data =
+				dbb->dbb_lock_mgr->readData2(lock->lck_type,
+											lock->getKeyPtr(), lock->lck_length,
+											lock->lck_owner_handle);
+			if (data != 0)
+			{
+				SLONG data2 = dbb->dbb_lock_mgr->readData(data);
+				if (data == data2 && data != 0 && data2 != 0)
+					skip = true;
+			}
+			if (!skip)
+			  count++;
+		}
 	}
```

Comment:
Additional call of LM in BTR_all looks like bad idea - it is not needed in 99% of cases and add performance overhead.
Also, i don't understand trick with storing request offset (lck_id) at lock data and double reading of it later in all places.
Also, here we have another leak of "Lock" instance.

Seems, you better should ask in fb-devel how to work with LM properly...